File: python/iplocation2.py
#!/usr/bin/env python
#coding:utf-8
# ip query by dat 
# power by ip138
import os
import socket
import mmap
import math
from struct import pack, unpack


class Ip138(object):

    def __init__(self, path):
        self.path = path
        self.db = None
        self.open()
        self.textOffset = 0
        self.total = 0
        self.init()


    def open(self):
        if not self.db:
            self.db = open(self.path, 'rb')
            self.db = mmap.mmap(self.db.fileno(), 0, access = 1)
        return self.db

    def init(self):
        self.db.seek(0)
        self.textOffset = unpack('I', self.db.read(4))[0]
        self.total = int((self.textOffset - 4 - 256*4) / 9)
        #print(self.textOffset)
        #print(self.total)

    def query(self, ip):
        try:
        	ip = unpack('!I', socket.inet_aton(ip))[0]
        except:
        	print('输入iP格式不对')
        	exit()

        left = 0
        right = 0
        first = ip>>24
        if first==0xff:
            self.db.seek(4+(first-1)*4)
            left = unpack('I', self.db.read(4))[0]+1
            right = self.total
        else:
            self.db.seek(4+first*4)
            left = unpack('I', self.db.read(4))[0]
            self.db.seek(4+(first+1)*4)
            right = unpack('I', self.db.read(4))[0]-1
            if right<1:
                right = self.total

        #print(first, left, right)
        i = 0
        while i<24:
            middle = int(math.floor((left+right)/2))
            if middle==left:
                ipOffset = int(4 + 256*4 + right*9)
                self.db.seek(ipOffset+4)
                textOffset = unpack('I', self.db.read(4))[0]
                self.db.seek(ipOffset+8)
                textLength = unpack('B', self.db.read(1))[0]
                #print(right, textOffset, textLength)

                self.db.seek(self.textOffset + textOffset)
                text = self.db.read(textLength)
                return text

            middleOffset = int(4 + 256*4 + middle*9)
            #print(left, right, middle)
            self.db.seek(middleOffset)
            middleIp = unpack('I', self.db.read(4))[0]
            if ip <= middleIp:
                right = middle
            else:
                left = middle

            i=i+1

        return "end"

    def __del__(self):
        if self.db:
            self.db.close()


def main():
    #ip = '61.139.2.69'
    #ip = '0.0.0.0'
    #ip = '255.255.255.255'
    #ip = '152.32.193.0'
    #ip = '1.0.2.1'
    ip = '110.81.112.0'
    dbpath = '../ip.dat'
    ipquery = Ip138(dbpath)
    a = ipquery.query(ip)
    print('%15s %s' % (ip, a))

if __name__ == '__main__':
    main()

File: java/IpLocation2.java
import java.io.*;
import java.nio.*; 

public class IpLocation2 {

    public static String PATH = "../ip.dat";
    private static IpLocation2 mInstance;
    public static IpLocation2 getInstance() {
        if (mInstance == null) {
            if(mInstance==null){
                mInstance = new IpLocation2();
            }
        }
        return mInstance;
    }
    private RandomAccessFile db;
    private long textOffset;
    private long total;

    public IpLocation2() {
        try {
            this.db = new RandomAccessFile(PATH, "r");

            this.textOffset = this.readIntLong(0);
            this.total = (this.textOffset - 4 - 256*4) / 9;
            //System.out.println(this.textOffset);
            //System.out.println(this.total);
        } catch (IOException ex) {
            ex.printStackTrace();
        }
    }

    public String findLocation(String strIP){
        long ip = ipToLong(strIP);
        //System.out.println(this.longToIp(ip));

        long first = ip >> 24;
        long left,right;
        if (first == 0xff) {
            left = this.readIntLong(4+(first-1)*4)+1;
            right = this.total;
        }else{
            left = this.readIntLong(4+first*4);
            right = this.readIntLong(4+(first+1)*4)-1;
            if(right<1){
                right = this.total;
            }
        }

        int i = 0;
        while(i<24){
            long middle = new Double(Math.floor((left+right)/2)).longValue();
            if(middle==left){
                long ipOffset = 4 + 256*4 + right*9;
                long textOffset = this.readIntLong(ipOffset + 4);
                long textLength = this.readByteLong(ipOffset + 8);

                return this.readString(this.textOffset + textOffset, textLength);
            }

            //System.out.printf("%d %d %d\n", left, right, middle);
            long middleOffset = 4 + 256*4 + middle*9;
            long middleIp = this.readIntLong(middleOffset);
            if(ip<=middleIp){
                right = middle;
            }else{
                left = middle;
            }

            i++;
        }


        return "fail";
    }

    private long readIntLong(long offset){
        try {
            this.db.seek(offset);
            byte[] buffer = new byte[4];
            this.db.read(buffer);
            return bytesToLong(buffer);
        } catch (IOException ex) {
            ex.printStackTrace();
        }
        return 0;
    }

    private long readByteLong(long offset){
        try {
            this.db.seek(offset);
            byte[] buffer = new byte[1];
            this.db.read(buffer);
            return buffer[0]&0xFF;
        } catch (IOException ex) {
            ex.printStackTrace();
        }
        return 0;
    }

    private String readString(long offset, long length){
        try {
            this.db.seek(new Long(offset).intValue());
            byte[] buffer = new byte[new Long(length).intValue()];
            this.db.read(buffer);
            return new String(buffer, "utf-8");
        } catch (IOException ex) {
            ex.printStackTrace();
        }
        return "";
    }

    private static int bytesToInt(byte[] b) {
        return (((b)[0] & 0xFF) | (((b)[1] << 8) & 0xFF00) | (((b)[2] << 16) & 0xFF0000) | (((b)[3] << 24) & 0xFF000000));
    }

    private static long bytesToLong(byte[] b) {
        ByteBuffer bb = ByteBuffer.wrap(new byte[] {0, 0, 0, 0, b[3], b[2], b[1], b[0]});
        return bb.getLong();
    }

    private static long ipToLong(String strIp) {
        long[] ip = new long[4];
        int position1 = strIp.indexOf(".");
        int position2 = strIp.indexOf(".", position1 + 1);
        int position3 = strIp.indexOf(".", position2 + 1);
        ip[0] = Long.parseLong(strIp.substring(0, position1));
        ip[1] = Long.parseLong(strIp.substring(position1 + 1, position2));
        ip[2] = Long.parseLong(strIp.substring(position2 + 1, position3));
        ip[3] = Long.parseLong(strIp.substring(position3 + 1));
        return ((ip[0] << 24) + (ip[1] << 16) + (ip[2] << 8) + ip[3]);
    }

    private static String longToIp(Long longIp){
        return ((longIp >> 24) & 0xFF) + "." +
                ((longIp >> 16) & 0xFF) + "." +
                ((longIp >> 8) & 0xFF) + "." +
                (longIp & 0xFF);
    }


}
